### PR TITLE
feat: two theorems about Real.rpow

### PR DESCRIPTION
**Pull request by casavaca**
---

Question: Do I need to add `_root_.Real.` prefix to theorem name? It's added similar theorems (such as `_root_.Real.finset_prod_rpow`) but I don't understand why.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by grunweg** [comment]
I can answer your question: it is related to namespacing. Roughly speaking (full documentation is here: ): if you're inside namespace `Name`
- if Lean seems a declaration (e.g., a theorem or definition) `foo`, it names this `Name.foo`.
- when you use a theorem `bar` in a proof and you're inside , Lean first checks if it knows a theorem `bar`, and if not, checks `Name.bar`
- if you want to refer to a theorem `Othername.bar`, you cannot simply write `Othername.bar` since Lean would look for a declaration `Name.Othername.bar`. Writing `_root_.Othername.bar` works; the `_root_` tells Lean to ignore open namespaces for "resolving" this declaration

Writing `namespace Name` gets you inside this namespace. Writing `open Name` means you can leave out the `Name.` prefix.

In this particular case, everything is fine. In `NNReal.lean`, you had a `namespace NNReal`, so needed to write `_root_.Real` to declare your theorems there. In this file, you're in the `Real` namespace already, so all is well.

File: Mathlib/Analysis/SpecialFunctions/Pow/Real.lean [hunk 1]
```diff
@@ -513,6 +513,31 @@ lemma rpow_mul_intCast (hx : 0 ≤ x) (y : ℝ) (n : ℤ) : x ^ (y * n) = (x ^ y
 /-! Note: lemmas about `(∏ i in s, f i ^ r)` such as `Real.finset_prod_rpow` are proved
 in `Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean` instead. -/
 
+/-- `rpow` version of `Finset.prod_pow_eq_pow_sum`. -/
+theorem finset_prod_rpow_eq_rpow_sum
+    {ι} (s : Finset ι) (f : ι → ℝ) (hs : ∀ i ∈ s, 0 ≤ f i) (r : ℝ) (hr : 0 ≤ r) :
```
Comment (grunweg) [inline Mathlib/Analysis/SpecialFunctions/Pow/Real.lean:518]: Two small suggestions:
- I think writing `{ι : Type*}` is preferred (that's certainly what the remaining file does)
- you can make some of your variables implicit, e.g. f or r. (They are mentioned by `hs` and `hr`.)
Altogether, I arrive at
```suggestion
    {ι : Type*} (s : Finset ι) {f : ι → ℝ} (hs : ∀ i ∈ s, 0 ≤ f i) {r : ℝ} (hr : 0 ≤ r) :

File: Mathlib/Analysis/SpecialFunctions/Pow/Real.lean [hunk 2]
```diff
@@ -513,6 +513,31 @@ lemma rpow_mul_intCast (hx : 0 ≤ x) (y : ℝ) (n : ℤ) : x ^ (y * n) = (x ^ y
...
+    ∏ i in s, r ^ f i = r ^ (∑ i in s, f i) := by
+  refine' Finset.cons_induction_on s
+    (p := fun s ↦ (∀ i ∈ s, 0 ≤ f i) → ∏ i in s, r ^ f i = r ^ ∑ i in s, f i) _ _ hs
```
Comment (grunweg) [inline Mathlib/Analysis/SpecialFunctions/Pow/Real.lean:521]: As I understand it, `refine` is preferred over `refine'`, especially in new code.
```suggestion
  refine Finset.cons_induction_on s
    (p := fun s ↦ (∀ i ∈ s, 0 ≤ f i) → ∏ i in s, r ^ f i = r ^ ∑ i in s, f i) ?_ ?_ hs

File: Mathlib/Analysis/SpecialFunctions/Pow/Real.lean [hunk 3]
```diff
@@ -513,6 +513,31 @@ lemma rpow_mul_intCast (hx : 0 ≤ x) (y : ℝ) (n : ℤ) : x ^ (y * n) = (x ^ y
...
+  · simp only [Finset.prod_empty, sum_empty, rpow_zero, implies_true]
+  · intros a s ha ih hs
+    have h : ∀ i ∈ s, 0 ≤ f i := fun i hi ↦ hs i (Finset.mem_cons.mpr (Or.inr hi))
+    rw [Finset.sum_cons, Finset.prod_cons, ih h, rpow_add_of_nonneg hr ?_ (Finset.sum_nonneg h)]
+    exact hs a (Finset.mem_cons.mpr (Or.inl rfl))
+
+theorem prod_div_rpow_distrib {ι} {s : Finset ι} {f g h : ι → ℝ} (hf : ∀ x ∈ s, 0 ≤ f x)
+    (hg : ∀ x ∈ s, 0 ≤ g x) :
```
Comment (grunweg) [inline Mathlib/Analysis/SpecialFunctions/Pow/Real.lean:529]: Same suggestion here, I'd write `{ι : Type*}` instead.

File: Mathlib/Analysis/SpecialFunctions/Pow/Real.lean [hunk 4]
```diff
@@ -513,6 +513,31 @@ lemma rpow_mul_intCast (hx : 0 ≤ x) (y : ℝ) (n : ℤ) : x ^ (y * n) = (x ^ y
...
+/-- `rpow` version of `Finset.prod_pow_eq_pow_sum`. -/
+theorem finset_prod_rpow_eq_rpow_sum
+    {ι} (s : Finset ι) (f : ι → ℝ) (hs : ∀ i ∈ s, 0 ≤ f i) (r : ℝ) (hr : 0 ≤ r) :
+    ∏ i in s, r ^ f i = r ^ (∑ i in s, f i) := by
+  refine' Finset.cons_induction_on s
+    (p := fun s ↦ (∀ i ∈ s, 0 ≤ f i) → ∏ i in s, r ^ f i = r ^ ∑ i in s, f i) _ _ hs
+  · simp only [Finset.prod_empty, sum_empty, rpow_zero, implies_true]
+  · intros a s ha ih hs
+    have h : ∀ i ∈ s, 0 ≤ f i := fun i hi ↦ hs i (Finset.mem_cons.mpr (Or.inr hi))
+    rw [Finset.sum_cons, Finset.prod_cons, ih h, rpow_add_of_nonneg hr ?_ (Finset.sum_nonneg h)]
```
Comment (grunweg) [inline Mathlib/Analysis/SpecialFunctions/Pow/Real.lean:525]: Just so you know: using dot notation, you can write this in a shorter way. (You don't have to, though!)
`rw [s.sum_cons, s.prod_cons, ih h, rpow_add_of_nonneg hr ?_ (s.sum_nonneg h)]`

**Review by grunweg — COMMENTED** [review]
Some minor comments; looks pretty good to me!

File: Mathlib/Analysis/SpecialFunctions/Pow/Real.lean (hunk 4 above)
Comment (casavaca) [reply Mathlib/Analysis/SpecialFunctions/Pow/Real.lean:525]: Thank you. Good to know.

File: Mathlib/Analysis/SpecialFunctions/Pow/Real.lean (hunk 3 above)
Comment (casavaca) [reply Mathlib/Analysis/SpecialFunctions/Pow/Real.lean]: Done.

**Review by grunweg — COMMENTED** [review]
So, locally your code looks good: I don't see anything that has to be improved.

However, thinking about it deeply, my intuition goes off and I wonder if there's a better *global* strategy. (`prod_pow_eq_pow_sum` has a one-line proof, for example)
Two ideas
- can this be deduced from Finset.prod_pow_eq_pow_sum?
- does mathlib have a `Complex` version of the lemma you're adding; can you deduce this from a complex version?

I don't know this part of the library too well, hence can only ask questions and not make good suggestions.

File: Mathlib/Analysis/SpecialFunctions/Pow/Real.lean [hunk 5]
```diff
@@ -513,6 +513,31 @@ lemma rpow_mul_intCast (hx : 0 ≤ x) (y : ℝ) (n : ℤ) : x ^ (y * n) = (x ^ y
 /-! Note: lemmas about `(∏ i in s, f i ^ r)` such as `Real.finset_prod_rpow` are proved
 in `Mathlib/Analysis/SpecialFunctions/Pow/NNReal.lean` instead. -/
 
+/-- `rpow` version of `Finset.prod_pow_eq_pow_sum`. -/
+theorem finset_prod_rpow_eq_rpow_sum
+    {ι : Type*} (s : Finset ι) {f : ι → ℝ} (hs : ∀ i ∈ s, 0 ≤ f i) {r : ℝ} (hr : 0 ≤ r) :
+    ∏ i in s, r ^ f i = r ^ (∑ i in s, f i) := by
```
Comment (j-loreaux) [inline Mathlib/Analysis/SpecialFunctions/Pow/Real.lean:519]: This already exists as [Real.rpow_sum_of_nonneg](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/SpecialFunctions/Pow/Real.html#Real.rpow_sum_of_nonneg)

File: Mathlib/Analysis/SpecialFunctions/Pow/Real.lean [hunk 6]
```diff
@@ -513,6 +513,31 @@ lemma rpow_mul_intCast (hx : 0 ≤ x) (y : ℝ) (n : ℤ) : x ^ (y * n) = (x ^ y
...
+    (p := fun s ↦ (∀ i ∈ s, 0 ≤ f i) → ∏ i in s, r ^ f i = r ^ ∑ i in s, f i) ?_ ?_ hs
+  · simp only [Finset.prod_empty, sum_empty, rpow_zero, implies_true]
+  · intros a s ha ih hs
+    have h : ∀ i ∈ s, 0 ≤ f i := fun i hi ↦ hs i (Finset.mem_cons.mpr (Or.inr hi))
+    rw [Finset.sum_cons, Finset.prod_cons, ih h, rpow_add_of_nonneg hr ?_ (Finset.sum_nonneg h)]
+    exact hs a (Finset.mem_cons.mpr (Or.inl rfl))
+
+theorem prod_div_rpow_distrib {ι : Type*} {s : Finset ι} {f g h : ι → ℝ} (hf : ∀ x ∈ s, 0 ≤ f x)
+    (hg : ∀ x ∈ s, 0 ≤ g x) :
+    ∏ x in s, (f x / g x) ^ (h x) = (∏ x in s, (f x) ^ (h x)) / (∏ x in s, (g x) ^ (h x)) := by
```
Comment (j-loreaux) [inline Mathlib/Analysis/SpecialFunctions/Pow/Real.lean:530]: This follows readily from [Real.finset_prod_rpow](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/SpecialFunctions/Pow/NNReal.html#Real.finset_prod_rpow). If you really want this lemma, you should put it in that file and use that theorem to prove it more simply.

**Comment by casavaca** [comment]
Closing this PR because they already exist. I need to become better at searching.

**Pull request closed without merging**